nexus/src/db/queries/network_interface.rs

```diff
@@ -642,6 +256,152 @@ fn decode_database_error(
+                .unwrap_or_else(|| {
+                    panic!("Unexpectedly small IPv4 subnetwork: '{}'", network)
+                })
+                .into()
+        }
+    }
+}
+
+/// The `NextGuestIpv4Address` query is a `NextItem` query for choosing the next
+/// available IPv4 address for a guest interface.
+#[derive(Debug, Clone, Copy)]
+pub struct NextGuestIpv4Address {
+    inner: NextItem<
+        db::schema::network_interface::table,
+        IpNetwork,
+        db::schema::network_interface::dsl::ip,
+        Uuid,
+        db::schema::network_interface::dsl::subnet_id,
+    >,
+}
+
+impl NextGuestIpv4Address {
+    pub fn new(subnet: Ipv4Network, subnet_id: Uuid) -> Self {
+        let subnet = IpNetwork::from(subnet);
+        let net = IpNetwork::from(first_available_address(&subnet));
+        let max_offset = last_address_offset(&subnet);
+        Self { inner: NextItem::new_scoped(net, subnet_id, max_offset) }
+    }
+}
+
+delegate_query_fragment_impl!(NextGuestIpv4Address);
+
+/// A `[NextItem`] subquery that selects the next empty slot for an interface.
+///
+/// Instances are limited to 8 interfaces (RFD 135). This pushes a subquery that
+/// looks like:
+///
+/// ```sql
+/// SELECT COALESCE((
+///     SELECT
+///         next_slot
+///     FROM
+///         generate_series(0, <max nics per instance>)
+///     AS
+///         next_slot
+///     LEFT OUTER JOIN
+///         network_interface
+///     ON
+///         (instance_id, time_deleted IS NULL, slot) =
+///         (<instance_id>, TRUE, next_slot)
+///     WHERE
+///         slot IS NULL
+///     LIMIT 1)
+/// ), 0)
+/// ```
+///
+/// That is, we select the lowest slot that has not yet been claimed by an
+/// interface on this instance, or zero if there is no such instance at all.
+///
+/// Errors
+/// ------
+///
+/// Note that the `generate_series` function is inclusive of its upper bound.
+/// We intentionally use the upper bound of the maximum number of NICs per
+/// instance. In the case where there are no available slots (the current max
+/// slot number is 7), this query will return 8. However, this violates the
+/// check on the slot column being between `[0, 8)`. This check violation is
+/// used to detect the case when there are no slots available.
+#[derive(Debug, Clone, Copy)]
+pub struct NextNicSlot {
+    inner: NextItem<
+        db::schema::network_interface::table,
+        i16,
+        db::schema::network_interface::dsl::slot,
+        Uuid,
+        db::schema::network_interface::dsl::instance_id,
+    >,
+}
+
+impl NextNicSlot {
+    pub fn new(instance_id: Uuid) -> Self {
+        Self {
+            inner: NextItem::new_scoped(0, instance_id, MAX_NICS_PER_INSTANCE),
+        }
+    }
+}
+
+impl QueryFragment<Pg> for NextNicSlot {
+    fn walk_ast<'a>(&'a self, mut out: AstPass<'_, 'a, Pg>) -> QueryResult<()> {
+        out.push_sql("SELECT COALESCE((");
+        self.inner.walk_ast(out.reborrow())?;
+        out.push_sql("), 0)");
```

Comment:
Isn't zero a valid slot? How would we distinguish allocation failure from "slot zero has been allocated"?

Reply:
Good question. Zero is indeed a valid slot. The `COALESCE` there is to handle the case where the `NextItem` query returns `NULL`. That only happens if there is no record in the table with the requested `instance_id`, i.e., there's not yet an interface for the instance (it's been verified to exist by the time we get here).

There is no "allocation failure," as I think you're imagining it. The `NextItem` query looks over the range `[0, 8]` for a slot. Note that it's inclusive on the upper bound, but that `8` is not a valid slot, `7` is the maximum allowed slot. So we'll return `8` if this instance is topped off in terms of interfaces, which will fail [this check](https://github.com/oxidecomputer/omicron/blob/a99069ac8967ea8aacc2724423e1b47fb30a9d7a/common/src/sql/dbinit.sql#L668) in the database. We detect that, and convert it to an error indicating that the instance does not support more interfaces.

Reply:
I think I'm understanding  - COALESCE catches the case where the instance doesn't exist, but the query was issued anyway.

If *that* happens, wouldn't this query return zero (risking colliding with the valid `[0, 7]` space of valid results)?

Reply:
It could be that I've messed up my logic, but here is my reasoning. `COALESCE(a, b)` returns `a` iff `a IS NOT NULL`, otherwise it returns `b`.

So, if the `NextItem` query returns `NULL`, we return `0`. The `NextItem` query looks like:

```sql
SELECT
    0 + offset AS slot
FROM
    generate_series(0, 8) AS slot
RIGHT OUTER JOIN
    network_interface
ON
    (instance_id, slot, time_deleted IS NULL) =
    (<instance_id>, 0 + offset, TRUE)
WHERE
    slot IS NULL
LIMIT 1
```

So, if there is an instance, then this returns the first value in the range `[0, 8]` where `network_interface.slot IS NULL`, since it's a right-outer join (return `NULL` for the left column where there is no match in the right part of the join-expression). For example, return `1` if there's an interface on the instance that's in slot `0`.

If the instance does not exist at all, then there are no rows in the subquery at all. That's because there are no records in `network_interface` with that `instance_id`. So the query is empty, i.e., it returns `NULL`. Thus the coalesce expression returns `0`.

In short, the _goal_ (again, I could have messed this up) is to return `0` if there is no interface at all for this instance (we're allocating the first), or `last_slot + 1` if we're allocating an _additional_ interface.

Reply:
I'm rereading your question. I think one part of the confusion is that the _instance_ does exist, it just doesn't have any interfaces. In that case, we _want_ to return zero here, since we're allocating the first.

Reply:
@smklein Let me know if this resolved your question. I'm planning to merge this, but we should make sure there's no other follow-up. Thanks!